### PR TITLE
Implement ServiceAccountCredentials.from_p12_keyfile_buffer().

### LINKED ISSUE #412: ServiceAccountCredentials should support directly passing in key file contents
(breaking out my comment from #401 as separate feature request)

Previously, `SignedJwtAssertionCredentials` allowed you to pass in the key file contents directly instead of passing in a file path (as shown [here](https://developers.google.com/api-client-library/python/auth/service-accounts)) It would be great if `ServiceAccountCredentials` also supported this, for both JSON and p12 keys.

In the meantime, this has prevented us from upgrading to v2.0.0, as our keys are pulled in dynamically at runtime and it's much easier for us to directly pass in the contents instead of having to first write to disk, pass in the path, and then do all the necessary cleanup.

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #412.

/cc @allenpc

### CODE REVIEW

File: tests/test_service_account.py [hunk 1]
```diff
@@ -173,6 +173,30 @@ def test_from_p12_keyfile_explicit(self):
...
+            # Seek back to the beginning so the buffer can be
+            # passed to the constructor.
+            file_obj.seek(0)
+            creds = ServiceAccountCredentials.from_p12_keyfile_buffer(
+                service_account_email, file_obj,
+                private_key_password=private_key_password,
+                scopes=scopes)
+        # Check the created object.
+        self.assertIsInstance(creds, ServiceAccountCredentials)
+        self.assertEqual(creds.client_id, None)
```
Comment (nathanielmanistaatgoogle) [inline tests/test_service_account.py]: Since this is already a `unittest2`-using file, `assertIsNone`, please (here and below).

Comment (dhermes) [inline tests/test_service_account.py]: @nathanielmanistaatgoogle PTAL

**Comment by allenpc** [comment]
Looking good from my side :+1:

**Comment by dhermes** [comment]
Couldn't decided between `from_p12_keyfile_buffer`, `from_p12_keyfile_stream` or something else. Any preference?

**Comment by nathanielmanistaatgoogle** [comment]
`from_p12_key_text`?

**Comment by dhermes** [comment]
It isn't text, it's a `read()`-able object

File: oauth2client/service_account.py [hunk 2]
```diff
@@ -248,6 +247,62 @@ def from_p12_keyfile(cls, service_account_email, filename,
...
+
+    @classmethod
+    def from_p12_keyfile_buffer(cls, service_account_email, file_buffer,
+                                private_key_password=None, scopes=''):
+        """Factory constructor from JSON keyfile.
+
+        Args:
+            service_account_email: string, The email associated with the
+                                   service account.
+            file_buffer: file-like object, Buffer that implements ``read()``
```
Comment (nathanielmanistaatgoogle) [inline oauth2client/service_account.py]: `file-like object that implements`.

Comment (dhermes) [inline oauth2client/service_account.py]: The style is `{type}, Sentence here that is a description`. You're saying `{type}` should be really long?

**Comment by nathanielmanistaatgoogle** [comment]
Drop `file` (without "-like"); the argument may have come from the filesystem or it may not have. `stream` might be as good as anything else; Python doesn't seem to have a consistent single term for what we're talking about although "file-like" is the most common that I've seen. `from_p12_key_stream`?

File: oauth2client/service_account.py (hunk 2 above)
Comment (nathanielmanistaatgoogle) [inline oauth2client/service_account.py]: Unfortunately, yes, I think so?

**Comment by dhermes** [comment]
Are you saying we should rename `from_p12_key_buffer` --> `from_p12_key_stream`?

**Comment by nathanielmanistaatgoogle** [comment]
I'm getting a little lost; let me back up a bit: for the methods that don't themselves read from the file system, why is it important that their users have the idea of a file called to mind when contemplating whether or not to use them?

**Comment by dhermes** [comment]
It's not important.

**Comment by nathanielmanistaatgoogle** [comment]
(I'm content to drop the discussion if I'm annoying you, by the way.)

I think `from_p12_key_buffer` and `from_p12_key_stream` are both fine as names?

File: oauth2client/service_account.py [hunk 3]
```diff
@@ -248,6 +247,62 @@ def from_p12_keyfile(cls, service_account_email, filename,
...
+    @classmethod
+    def from_p12_keyfile_buffer(cls, service_account_email, file_buffer,
+                                private_key_password=None, scopes=''):
+        """Factory constructor from JSON keyfile.
+
+        Args:
+            service_account_email: string, The email associated with the
+                                   service account.
+            file_buffer: stream, A buffer that implements ``read()``
+                         and contains the PKCS#12 key contents.
```
Comment (dhermes) [inline oauth2client/service_account.py:287]: @nathanielmanistaatgoogle I updated this. PTAL

**Comment by nathanielmanistaatgoogle** [comment]
Looks good; squash commits and merge at your leisure. :+1:

**Pull request merged**